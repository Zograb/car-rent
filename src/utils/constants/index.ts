export * from './enums';
export * from './tailwindConfig';
