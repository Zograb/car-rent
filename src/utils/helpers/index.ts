export * from './formatNumberToCurrency';
export * from './getPercentageDifference';
export * from './getRandomColor';
export * from './getCurrentBreakpoint';
