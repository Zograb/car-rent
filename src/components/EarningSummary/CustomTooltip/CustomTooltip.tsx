import type { TooltipProps } from 'recharts';
import { useCallback, useMemo } from 'react';
import { format, subYears } from 'date-fns';
import 'twin.macro';

// Utils
import { formatNumberToCurrency } from '@/utils';

const CustomTooltip = ({ active, payload, label }: TooltipProps<number, string>) => {
  const formatEarningValue = useCallback((value?: number) => {
    if (value) {
      return formatNumberToCurrency(
        value,
        {
          notation: 'compact'
        },
      );
    }
    return 0;
  }, []);

  const value = useMemo(() => {
    if (payload?.length) {
      return payload.find((item) => item.dataKey === 'value');
    }
    return null;
  }, [payload]);

  if (active && value) {
    return (
      <div tw="bg-white/80 shadow rounded-sm p-2" data-testid="custom-tooltip">
        <p tw="text-lg">{format(label, 'MMMM')}</p>
        <p tw="font-light">
          {`${format(label, 'yyyy')}: ${formatEarningValue(value.value)}`}
        </p>
        <p tw="font-light">
          {`${format(subYears(label, 1), 'yyyy')}: ${
            formatEarningValue(value.payload?.previousYearValue)
          }`}
        </p>
      </div>
    );
  }

  return null;
};

export default CustomTooltip;
