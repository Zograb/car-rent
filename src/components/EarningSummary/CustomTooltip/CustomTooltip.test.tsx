import { render } from '@testing-library/react';
import { format, subYears } from 'date-fns';

// Components
import CustomTooltip from './CustomTooltip';

// Utils
import { formatNumberToCurrency } from '@/utils';


const payload = [
  {
    value: 10000,
    dataKey: 'value',
    payload: {
      previousYearValue: 15000,
    },
  },
]

describe('CustomTooltip', () => {
  // Temporarily workaround for bug in @testing-library/react when use user-event with `vi.useFakeTimers()`
  beforeAll(() => {
    const _jest = globalThis.jest;
    globalThis.jest = {
      ...globalThis.jest,
      advanceTimersByTime: jest.advanceTimersByTime,
    };
    return () => void (globalThis.jest = _jest);
  });

  beforeEach(() => {
    jest.useFakeTimers();
    jest.setSystemTime(new Date(2000, 12, 12));
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  it('should match snapshot', () => {
    const { container } = render(
      <CustomTooltip
        label={new Date()}
        payload={payload}
        active
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('should render custom tooltip', () => {
    const { getByTestId } = render(
      <CustomTooltip
        label={new Date()}
        payload={payload}
        active
      />
    );
    const tooltip = getByTestId('custom-tooltip');
    expect(tooltip).toBeInTheDocument();
  });

  it('should render label and value in tooltip', () => {
    const date = new Date();
    const { getByText } = render(
      <CustomTooltip
        label={date}
        payload={payload}
        active
      />
    );
    const label = getByText(format(date, 'MMMM'));
    const value1 = formatNumberToCurrency(10000, {
      notation: 'compact'
    });
    const value2 = formatNumberToCurrency(15000, {
      notation: 'compact',
    });
    expect(label).toBeInTheDocument();
    expect(getByText(`${format(date, 'yyyy')}: ${value1}`)).toBeInTheDocument();
    expect(getByText(`${format(subYears(date, 1), 'yyyy')}: ${value2}`)).toBeInTheDocument();
  });

});
