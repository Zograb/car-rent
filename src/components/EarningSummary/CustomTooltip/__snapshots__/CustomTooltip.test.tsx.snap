// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomTooltip should match snapshot 1`] = `
.c0 {
  border-radius: 0.125rem;
  background-color: rgba(255, 255, 255, 0.8);
  padding: 0.5rem;
  --tw-shadow: 0 1px 3px 0 rgb(0 0 0 / 0.1),0 1px 2px -1px rgb(0 0 0 / 0.1);
  --tw-shadow-colored: 0 1px 3px 0 var(--tw-shadow-color),0 1px 2px -1px var(--tw-shadow-color);
  box-shadow: var(--tw-ring-offset-shadow, 0 0 #0000),var(--tw-ring-shadow, 0 0 #0000),var(--tw-shadow);
}

.c1 {
  font-size: 1.125rem;
  line-height: 1.75rem;
}

.c2 {
  font-weight: 300;
}

.c3 {
  font-weight: 300;
}

<div>
  <div
    class="c0"
    data-testid="custom-tooltip"
  >
    <p
      class="c1"
    >
      January
    </p>
    <p
      class="c2"
    >
      2001: $10K
    </p>
    <p
      class="c3"
    >
      2000: $15K
    </p>
  </div>
</div>
`;
